### DISCUSSION TITLE
Unable to serve a slug file for a route with a query string

### DISCUSSION THREAD

**Discussion by alex-r89** [vercel/next.js#48358] [Help]
### Summary

Hi,

I am unsure if this is actually a bug/unsupported feature or something I am doing wrong. To explain the issue:

- I have an application with a search functionality via a search page
- This sits under the `/search` slug - to clarify that would be `http://localhost:3000/search`
- This is served from a file located inside `pages/search/index.js`
- When a user searches from this page, I want them to be taken to this url for the results:  `http://localhost:3000/search/?q=search_term`
- This url should be served from a separate file, that being `pages/search/[slug].js`
- Once a user is on that route and has been served that slug file, **they have the ability to search again.**

It seems like nextJS will not take the user to the [slug].js file and instead renders the `index.js` file when a users visits the url `http://localhost:3000/search/?q=search_term`.

**To clarify:** When a user visits `http://localhost:3000/search/?q=search_term` the slug file (`/pages/search/[slug].js`) is not rendered, the index file (`/pages/search/index.js`) is - which is not what I want.

- I have tried this with both `[slug].jsx` and `[...slug].js` but neither works.
- I have tried naming the slug file `[[...slug]].js` but this is not allowed as it conflicts with the `index.js` file inside the folder `pages/search/index.js`
- I have also tried enabling the Trailing Slash feature in the next config, which didnt work
- I have also tried moving the file which renders the /search url into the root of pages, and naming it `search.js` - so this would be `pages/search.js` but this also didnt work

There seems to be no way for me to serve one file for the  `http://localhost:3000/search/?q=search_term` route, and a separate file for the `http://localhost:3000/search` route

Is this not supported by NextJS or am i doing something wrong? I thought it would be quite a common setup for people to use, especially when users often want to search again from a results page, so it seems odd that a query string causes the wrong file to be served?

**Comment by 808vita** [comment]
Hey Alex , "http://localhost:3000/search/?q=search_term" will take to "/pages/search/index.js" that is the expected correct behavior . You would be able to access the "q" as query params.

To able to navigate to "pages/search/[slug].js" page you would need to place like "http://localhost:3000/search/search_term" . With this you would be get hold of the searched term by using params .Make use of Router from "next/router"; it would hold  params:{slug:"search_term"} .


I did try to make search recently will attach links to relevant files ( but slightly different way) , hope it helps you .
Links:
[search page](https://github.com/808vita/nextjs-blog-with-auth/blob/deploy/src/pages/search/index.tsx) 

[search component](https://github.com/808vita/nextjs-blog-with-auth/blob/deploy/src/components/searchComponent/searchComponent.tsx) 

[client fetch api call function](https://github.com/808vita/nextjs-blog-with-auth/blob/deploy/src/fetchApiCalls/fetchApiCalls.ts) 

[next api file ](https://github.com/808vita/nextjs-blog-with-auth/blob/deploy/src/pages/api/blog-search/get.ts) 

[apiSearchBlogPost](https://github.com/808vita/nextjs-blog-with-auth/blob/deploy/apiControllers/blogSearch/apiHandlerController.ts) 

[Try out the search deployment here ](https://nextjs-blog-with-auth.vercel.app/search) 
![image](https://user-images.githubusercontent.com/97225946/231874644-e54b398b-04f9-432c-84e1-41c7d80070d6.png)

And also thanks for stating clearly what you were trying to do & experiencing , it made things a lot easier

**Reply by 808vita** [reply]
if you could post code snippets, it would also help .

**Reply by alex-r89** [reply]
Hi,

Thanks for the reply. 

So there is no way to go to `http://localhost:3000/search/?q=search_term` and serve the slug file?

This seems odd because we may want the root page, ie `/search` to contain different information to the results page yet still all be under the same slug?

For example, a user may go to `/search`, be instructed on how to search. Then perform a search and be taken to a results page under `/search/?q=search_term`, with the query string in the url - which is served by the slug. This doesn't seem abnormal to me and I would have thought this is normal?

Please see the repo below highlighting my issue: 

https://github.com/alex-r89/route-param-issue-nextjs

Note, Ive thrown this together as quickly as possible, this is by no means good quality - it just shows the issue im attempting to solve.

I may need to do what you suggest and navigate to `/search/search_term` instead though to solve this.

**Reply by riceboyler** [reply]
The thing is, `?q=search_term` is not a slug, at least as far as the Next Router is concerned. That's where the problem comes in. `search/?q=search_term` is seen as `search/index` with a query parameter of `search_term`. If you make it be `/search/search_term` it should work as you would expect it to.

**Reply by 808vita** [reply]
There are several ways to tackle this , I don't have context of the requirements .

option 1 - when user hits `/search/?q=search_term` route them to `/search/results?q=search_term`  or `/search/search_term` 
 
option 2 - directly hit  `/search/results?q=search_term` 
(**call the path as results , s , q  or place the search term itself or anything) 
Either way you would have the access to the query params. I would prefer the second option.

option 3 -conditional rendering . check if url has query values , if it exists just show the results JSX .otherwise show the search instruction .
With this you don't need another page , just split it into two components and render conditionally .

**Comment by 808vita** [comment]
Not tested but this is option 3

```
import React, { useState } from "react";
import { useRouter } from "next/router";

const SearchPage = () => {
  const [searchTerm, setSearchTerm] = useState("");
  const router = useRouter();

  const handleSearch = () => {
    router.push({
      pathname: `/search`,
      query: { q: searchTerm },
    });
  };

  const { q } = router.query;

  return (
    <>
      {q ? (
        <div>{`You searched for ${q}`}</div>
      ) : (
        <form>
          <input
            type="text"
            placeholder="Search.."
            name="q"
            onChange={(event) => setSearchTerm(event.target.value)}
          />
          <button onClick={handleSearch}>Submit</button>
        </form>
      )}
    </>
  );
};

export default SearchPage;

```

**Reply by 808vita** [reply]
tested and this does work 
![image](https://user-images.githubusercontent.com/97225946/231897070-5e4ab097-b232-40a7-b017-4e8114500adb.png)

![image](https://user-images.githubusercontent.com/97225946/231897174-c7a3105d-b3d7-442c-b4f5-79f61426d58c.png)

**Reply by 808vita** [reply]
```
  const handleSearch = (e) => {
    e.preventDefault();
    router.push({
      pathname: `/search`,
      query: { q: searchTerm },
    });
  };
```
make sure to add e.preventDefault(); otherwise refresh would occur.
